kani-compiler/src/codegen_boogie/compiler_interface.rs

```diff
@@ -0,0 +1,666 @@
+                        MonoItem::GlobalAsm(_) => {} // We have already warned above
+                    }
+                }
+            },
+            "codegen",
+        );
+
+        (bcx, items)
+    }
+}
+
+impl CodegenBackend for BoogieCodegenBackend {
+    fn metadata_loader(&self) -> Box<MetadataLoaderDyn> {
+        Box::new(rustc_codegen_ssa::back::metadata::DefaultMetadataLoader)
+    }
+
+    fn provide(&self, providers: &mut Providers) {
+        provide::provide(providers, &self.queries.lock().unwrap());
+    }
+
+    fn provide_extern(&self, providers: &mut ty::query::ExternProviders) {
+        provide::provide_extern(providers);
+    }
+
+    fn print_version(&self) {
+        println!("Kani-boogie version: {}", env!("CARGO_PKG_VERSION"));
+    }
+
+    fn locale_resource(&self) -> &'static str {
+        // We don't currently support multiple languages.
+        DEFAULT_LOCALE_RESOURCE
+    }
+
+    fn codegen_crate(
+        &self,
+        tcx: TyCtxt,
+        rustc_metadata: EncodedMetadata,
+        _need_metadata_module: bool,
+    ) -> Box<dyn Any> {
+        // Queries shouldn't change today once codegen starts.
+        let queries = self.queries.lock().unwrap().clone();
+        check_target(tcx.sess);
+        check_options(tcx.sess);
+        check_crate_items(tcx, queries.ignore_global_asm);
+
+        // Codegen all items that need to be processed according to the selected reachability mode:
+        //
+        // - Harnesses: Generate one model per local harnesses (marked with `kani::proof` attribute).
+        // - Tests: Generate one model per test harnesses.
+        // - PubFns: Generate code for all reachable logic starting from the local public functions.
+        // - None: Don't generate code. This is used to compile dependencies.
+        let base_filename = tcx.output_filenames(()).output_path(OutputType::Object);
+        let reachability = queries.reachability_analysis;
+        let mut results = BoogieCodegenResults::new(tcx, reachability);
+        match reachability {
+            ReachabilityType::Harnesses => {
+                // Cross-crate collecting of all items that are reachable from the crate harnesses.
+                let harnesses = queries.target_harnesses();
+                let mut items: HashSet<DefPathHash> = HashSet::with_capacity(harnesses.len());
+                items.extend(harnesses.into_iter());
+                let harnesses =
+                    filter_crate_items(tcx, |_, def_id| items.contains(&tcx.def_path_hash(def_id)));
+                for harness in harnesses {
+                    let model_path =
+                        queries.harness_model_path(&tcx.def_path_hash(harness.def_id())).unwrap();
+                    let (gcx, items) =
+                        self.codegen_items(tcx, &[harness], model_path, &results.machine_model);
+                    results.extend(gcx, items, None);
+                }
+            }
+            ReachabilityType::Tests => {
```

Comment:
FYI, this code will change quite a bit soon... I'll see if I can factor it out so we don't have to replicate it around here.